File: Arrays and Strings/isPermutation.cxx
//
//  Cracking the Coding Interview Q1.2 
//  Test if two given strings are valid
//  permutations of eachother
//
//  Joel Rorseth
//  Created on May 1, 2017
//

#include <iostream>

// Solution
bool isPermutation(const std::string & s, const std::string & t) {
    
    if (s.length() != t.length())
        return false;
    
    // Makes copies to avoid changing passed strings
    std::string s1 = s;
    std::string s2 = t;

    // Sort characters alphabetically
    std::sort(s1.begin(), s1.end());
    std::sort(s2.begin(), s2.end());

    // If identical, then valid permutation!
    return s1 == s2;
}

// Demonstration
int main(int argc, char * argv[]) {

    if (argc < 3) {
        std::cout << "Usage: " << argv[0] << " <string1> <string2>" << std::endl;
        exit(0);
    }

    auto result = isPermutation(argv[1], argv[2]);

    if (result)
        std::cout << argv[1] << " and " << argv[2] 
            << " are valid permutations of eachother" << std::endl;
    else 
        std::cout << argv[1] << " and " << argv[2]
            << " are not permutations of eachother" << std::endl;
}

File: Arrays and Strings/Python/is_permutation.py
#
# Are two strings permutations of eachother?
# Joel Rorseth
#

# Q: Case sensitive? YES
# Q: Is whitespace significant? YES
# Solution: Compare length, then make set of chars in a, compare with b
# Complexity: O(n)


def is_permutation(a, b):

    # Nice, but does extra work (checks whole string) if non-permutation
    #return False if (len(a) != len(b)) else set(a)==set(b)

    if (len(a) != len(b)):
        return False

    # Make counter for all ASCII characters, count chars in 'a'
    a_chars = [0] * 128
    for c in a:
        a_chars[ord(c)] += 1

    # Negative counts will immediately be false
    for c in b:
        a_chars[ord(c)] -= 1

        if a_chars[ord(c)] < 0:
            return False

    return True


print("dog, cat")
print(is_permutation("dog", "cat"))

print("\nabcd, bdca")
print(is_permutation("abcd", "bdca"))
